codex-rs/execpolicy2/src/policy.rs

```diff
@@ -0,0 +1,95 @@
+use crate::decision::Decision;
+use crate::error::Error;
+use crate::error::Result;
+use crate::rule::Rule;
+use crate::rule::RuleMatch;
+use multimap::MultiMap;
+use serde::Deserialize;
+use serde::Serialize;
+use shlex::try_join;
+
+#[derive(Clone, Debug)]
+pub struct Policy {
+    rules_by_program: MultiMap<String, Rule>,
+}
+
+impl Policy {
+    pub fn new(rules_by_program: MultiMap<String, Rule>) -> Self {
+        Self { rules_by_program }
+    }
+
+    pub fn rules(&self) -> &MultiMap<String, Rule> {
+        &self.rules_by_program
+    }
+
+    pub fn check(&self, cmd: &[String]) -> Evaluation {
+        let rules = match cmd.first() {
+            Some(first) => match self.rules_by_program.get_vec(first) {
+                Some(rules) => rules,
+                None => return Evaluation::NoMatch,
+            },
+            None => return Evaluation::NoMatch,
+        };
+
+        let matched_rules: Vec<RuleMatch> =
+            rules.iter().filter_map(|rule| rule.matches(cmd)).collect();
+        match matched_rules.iter().map(RuleMatch::decision).max() {
+            Some(decision) => Evaluation::Match {
+                decision,
+                matched_rules,
+            },
+            None => Evaluation::NoMatch,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Evaluation {
+    NoMatch,
+    Match {
+        decision: Decision,
+        matched_rules: Vec<RuleMatch>,
+    },
+}
+
+impl Evaluation {
+    pub fn is_match(&self) -> bool {
+        matches!(self, Self::Match { .. })
+    }
+}
+
+/// Count how many rules match each provided example and error if any example is unmatched.
+pub(crate) fn validate_match_examples(rules: &[Rule], matches: &[Vec<String>]) -> Result<()> {
+    let match_counts = rules.iter().fold(vec![0; matches.len()], |counts, rule| {
```

Comment:
I know I pushed you to favor `iter()`, but I always raise an eyebrow when I see `fold()` used.

It feels more direct to loop over matches and on each iteration of the loop, check if any of `rules` matches the `match`.

Just allocate a `mut Vec` that you add to for `unmatched_examples`.